programs/fileio_asyncio.h

```diff
@@ -113,6 +128,44 @@ WritePoolCtx_t* AIO_WritePool_create(FIO_prefs_t* const prefs, size_t bufferSize
  * Frees and releases a writePool and its resources. Closes destination file. */
 void AIO_WritePool_free(WritePoolCtx_t* ctx);
 
+/* AIO_ReadPool_create:
+ * Allocates and sets and a new readPool including its included jobs.
+ * bufferSize should be set to the maximal buffer we want to read at a time, will also be used
+ * as our basic read size. */
+ReadPoolCtx_t* AIO_ReadPool_create(FIO_prefs_t* const prefs, size_t bufferSize);
+
+/* AIO_ReadPool_free:
+ * Frees and releases a readPool and its resources. Closes source file. */
+void AIO_ReadPool_free(ReadPoolCtx_t* ctx);
+
+/* AIO_ReadPool_consumeBytes:
+ * Consumes byes from srcBuffer's beginning and updates srcBufferLoaded accordingly. */
+void AIO_ReadPool_consumeBytes(ReadPoolCtx_t *ctx, size_t n);
+
+/* AIO_ReadPool_fillBuffer:
+ * Makes sure buffer has at least n bytes loaded (as long as n is not bigger than the initalized bufferSize).
+ * Returns if srcBuffer has at least n bytes loaded or if we've reached the end of the file.
+ * Return value is the number of bytes added to the buffer.
+ * Note that srcBuffer might have up to 2 times bufferSize bytes. */
+size_t AIO_ReadPool_fillBuffer(ReadPoolCtx_t *ctx, size_t n);
```

Comment:
We may be able to simplify this a little bit by saying you MUST consume the entire source buffer. And instead of storing it in the `ReadPoolCtx_t`, return the buffer from the `AIO_ReadPool_fillBuffer()` function.

Maybe that way we could avoid some memcpy's, or just simplify the API.

Reply:
The problem here is that you might have multiple frames in a single file, and those frames can theoretically be different formats as well. That means that we can't assume the whole buffer is consumed.

A possible API here would be `consume` and `read` where `read` would return a pointer to a buffer and a buffer size. That'd be cleaner as we won't expose `->srcBuffer` to the user.
The downsize is that kind of change would require more changes to `fileio.c`. As this is a big patch already I elected to avoid those, but willing to rethink it.

Reply:
> The problem here is that you might have multiple frames in a single file, and those frames can theoretically be different formats as well. That means that we can't assume the whole buffer is consumed.

Good point!

> A possible API here would be consume and read where read would return a pointer to a buffer and a buffer size. That'd be cleaner as we won't expose ->srcBuffer to the user.
> The downsize is that kind of change would require more changes to fileio.c. As this is a big patch already I elected to avoid those, but willing to rethink it.

I have a slight preference for returning a buffer pointer. That might also allow us to avoid some `memcpy()`. If we can give pointers directly to buffers that the jobs read into.

E.g. have one extra input buffer. And when we need to read the next job, swap the job's buffer with the current input buffer. That way the job can continue with its new buffer, and the input can read directly from the job's old buffer w/o a memcpy.

I'm sure theres different ways to achieve that as well, and maybe those are cleaner. Just a thought.